https://huggingface.co/docs/text-embeddings-inference/main/en/local_cpu#step-1-install-rust

In [1]:
# 1️⃣ الانتقال إلى مجلد العمل في Google Colab
%cd /content

# 2️⃣ تثبيت Rust في Google Colab
!curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y

# تحديث المسار بعد التثبيت
import os
os.environ["PATH"] += ":/root/.cargo/bin"

# التحقق من تثبيت Rust و Cargo
!rustc --version
!cargo --version

/content
info: downloading installer
info: profile set to 'default'
info: default host triple is x86_64-unknown-linux-gnu
info: syncing channel updates for 'stable-x86_64-unknown-linux-gnu'
info: latest update on 2025-01-30, rust version 1.84.1 (e71f9a9a9 2025-01-27)
info: downloading component 'cargo'
info: downloading component 'clippy'
info: downloading component 'rust-docs'
info: downloading component 'rust-std'
info: downloading component 'rustc'
 70.6 MiB /  70.6 MiB (100 %)  33.3 MiB/s in  2s ETA:  0s
info: downloading component 'rustfmt'
info: installing component 'cargo'
  8.8 MiB /   8.8 MiB (100 %)   8.0 MiB/s in  1s ETA:  0s
info: installing component 'clippy'
info: installing component 'rust-docs'
 16.5 MiB /  16.5 MiB (100 %)   1.7 MiB/s in  7s ETA:  0s
info: installing component 'rust-std'
 26.7 MiB /  26.7 MiB (100 %)   6.5 MiB/s in  4s ETA:  0s
info: installing component 'rustc'
 70.6 MiB /  70.6 MiB (100 %)   7.7 MiB/s in  9s ETA:  0s
info: installing component 'rustf

In [2]:
!cargo new my_rust_project
%cd my_rust_project

    Creating binary (application) `my_rust_project` package
note: see more `Cargo.toml` keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
/content/my_rust_project


In [3]:
!cargo build
!cargo run

   Compiling my_rust_project v0.1.0 (/content/my_rust_project)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.54s
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
     Running `target/debug/my_rust_project`
Hello, world!


In [4]:
!cargo new my_rust_project2
%cd my_rust_project2

    Creating binary (application) `my_rust_project2` package
note: see more `Cargo.toml` keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
/content/my_rust_project/my_rust_project2


In [5]:
%%writefile src/main.rs

fn main() {
    println!("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
}

Overwriting src/main.rs


In [6]:
!cargo build
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.24s
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.01s
     Running `target/debug/my_rust_project2`
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa


In [ ]:
use reqwest::blocking::Client;
use serde_json::json;

fn main() {
    let api_token = "your_huggingface_api_token"; // ضع توكن المصادقة الخاص بك هنا
    let url = "https://api-inference.huggingface.co/models/gpt2";

    let client = Client::new();
    let request_body = json!({ "inputs": "Hello, world!" });

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}", api_token))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send();

    match response {
        Ok(resp) => match resp.text() {
            Ok(text) => println!("Response: {}", text),
            Err(err) => eprintln!("Error reading response text: {}", err),
        },
        Err(err) => eprintln!("Error sending request: {}", err),
    }
}


In [ ]:
%%writefile src/main.rs

fn main() {
    println!("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa");
}

In [ ]:
!cargo build
!cargo run

In [9]:
%cd /content/my_rust_project/my_rust_project2

/content/my_rust_project/my_rust_project2


In [12]:
%%writefile main.rs
use reqwest::blocking::Client;
use serde_json::json;

fn main() {
    let api_token = "xxxxxxxxxxxxxx"; // استبدل بالتوكن الخاص بك
    let url = "https://api-inference.huggingface.co/models/gpt2";

    let client = Client::new();
    let payload = json!({ "inputs": "Hello, world!" });

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}", api_token))
        .header("Content-Type", "application/json")
        .json(&payload)
        .send();

    match response {
        Ok(res) => match res.text() {
            Ok(text) => println!("Response: {}", text),
            Err(err) => eprintln!("Error reading response: {}", err),
        },
        Err(err) => eprintln!("Request failed: {}", err),
    }
}



Overwriting main.rs


In [13]:
!cargo build
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
error[E0433]: failed to resolve: use of undeclared crate or module `reqwest`
 --> src/main.rs:1:5
  |
1 | use reqwest::blocking::Client;
  |     ^^^^^^^ use of undeclared crate or module `reqwest`

error[E0432]: unresolved import `serde_json`
 --> src/main.rs:2:5
  |
2 | use serde_json::json;
  |     ^^^^^^^^^^ use of undeclared crate or module `serde_json`

Some errors have detailed explanations: E0432, E0433.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `my_rust_project2` (bin "my_rust_project2") due to 2 previous errors
   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
error[E0433]: failed to resolve: use of undeclared crate or module `reqwest`
 --> src/main.rs:1:5
  |
1 | use reqwest::blocking::Client;
  |     ^^^^^^^ use of undeclared crate or module `reqwest`

error[E0432]: unresolved import `serde_json`
 --> src/main.rs:2:5

In [14]:
!cargo install --path router -F mkl

error: `/content/my_rust_project/my_rust_project2/router` is not a directory. --path must point to a directory containing a Cargo.toml file.


In [ ]:
model=BAAI/bge-large-en-v1.5
revision=refs/pr/5

text-embeddings-router --model-id $model --revision $revision --port 8080

In [15]:
!pip install -q aiohttp==3.8.3 datasets==2.14.6 pandas==1.5.3 requests==2.31.0 tqdm==4.66.1 huggingface-hub>=0.20

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 1.5.3 which is incompatible.
google-colab 1.0.0 requires requests==2.32.3, but you have requests 2.31.0 which is incompatible.
cudf-cu12 24.12.0 requires pandas<2.2.4dev0,>=2.0, but you have pandas 1.5.3 which is incompatible.
xarray 2025.1.2 requires pandas>=2.1, but you have pandas 1.5.3 which is incompatible.
torch 2.5.1+cu124 requires nvidia-cublas-cu12==12.4.5.8; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cublas-cu12 12.5.3.2 which is incompatible.
torch 2.5.1+cu124 requires nvidia-cuda-cupti-cu12==12.4.127; platform_system == "Linux" and platform_machine == "x86_64", but you have nvidia-cuda-cupti-cu12 12.5.82 which is incompatible.
torch 2.5.1+cu124 requires nvidia-cuda-nvrtc-cu12==12.4.127; platform_system 

In [16]:
!sudo apt-get install libssl-dev gcc -y

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
gcc is already the newest version (4:11.2.0-1ubuntu1).
gcc set to manually installed.
libssl-dev is already the newest version (3.0.2-0ubuntu1.18).
0 upgraded, 0 newly installed, 0 to remove and 20 not upgraded.


https://huggingface.co/docs/tokenizers/main/en/quicktour?code=rust

https://doc.rust-lang.org/rust-by-example/

https://doc.rust-lang.org/rust-by-example/

In [ ]:
cargo update

cargo run



use reqwest::blocking::Client;
use serde_json::json;

fn main() {
    let api_token = "XXXXXXXXXXXXXXXX"; // ضع توكن المصادقة الخاص بك هنا
    let url = "https://api-inference.huggingface.co/models/gpt2";

    let client = Client::new();
    let request_body = json!({ "inputs": "Hello, world!" });

    let response = client
        .post(url)
        .header("Authorization", format!("Bearer {}", api_token))
        .header("Content-Type", "application/json")
        .json(&request_body)
        .send();

    match response {
        Ok(resp) => match resp.text() {
            Ok(text) => println!("Response: {}", text),
            Err(err) => eprintln!("Error reading response text: {}", err),
        },
        Err(err) => eprintln!("Error sending request: {}", err),
    }
}

In [23]:
%cd /content/my_rust_project/my_rust_project2

%cd /content/my_rust_project/my_rust_project2

%%writefile src/main.rs

{
    println!("qqqqqqqqqqqqqqqqqqqqqqqq");
}







SyntaxError: invalid syntax (<ipython-input-23-91f1192099b0>, line 8)

@@@@@@@@@@@@@@@@@@@@@@@@@@

In [24]:
%%writefile src/main.rs
fn main() {
    println!("مرحباً بالعالم من راست!");
}

Writing src/main.rs


In [25]:
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.44s
     Running `target/debug/my_rust_project2`
مرحباً بالعالم من راست!


#####################################

In [27]:

%%writefile src/main.rs
fn main() {
    println!("مرحباً");
}




Overwriting src/main.rs


In [28]:
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.32s
     Running `target/debug/my_rust_project2`
مرحباً


aشغال

In [29]:
%%writefile src/main.rs
use std::collections::HashMap;
use std::io::{self, Write};

// هنا يتم تعريف "النموذج" اللغوي البسيط كقاموس (HashMap)
// في الواقع، النماذج اللغوية الحقيقية أكثر تعقيدًا بكثير
fn load_model() -> HashMap<&'static str, &'static str> {
    let mut model = HashMap::new();
    model.insert("مرحباً", "مرحباً بك!");
    model.insert("كيف حالك؟", "أنا بخير، شكراً لسؤالك.");
    model.insert("ما اسمك؟", "أنا نموذج لغوي بسيط في Rust.");
    model.insert("وداعاً", "إلى اللقاء!");
    model
}

// دالة الاستدلال الأساسية
fn infer(model: &HashMap<&'static str, &'static str>, input: &str) -> String {
    match model.get(input) {
        Some(response) => response.to_string(), // إذا وجدنا إجابة في النموذج
        None => "لا أفهم ما تقصده.".to_string(), // إذا لم نجد إجابة
    }
}

fn main() {
    let model = load_model(); // تحميل "النموذج"

    println!("نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):");

    loop {
        print!("أنت: ");
        io::stdout().flush().unwrap(); // لضمان ظهور "أنت:" قبل الإدخال

        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap(); // قراءة سطر الإدخال من المستخدم

        let input = input.trim(); // إزالة المسافات البيضاء الزائدة في البداية والنهاية

        if input == "وداعاً" {
            println!("النموذج: إلى اللقاء!");
            break; // الخروج من الحلقة إذا قال المستخدم "وداعاً"
        }

        let response = infer(&model, input); // إجراء الاستدلال باستخدام النموذج والمدخل
        println!("النموذج: {}", response); // طباعة الاستجابة
    }
}

Overwriting src/main.rs


In [30]:
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1.17s
     Running `target/debug/my_rust_project2`
نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):
أنت: hi
النموذج: لا أفهم ما تقصده.
أنت: ^C


In [31]:
%%writefile src/main.rs
// **تنبيه: هذا الكود مبسط جدًا وغير مكتمل. يحتاج إلى تفاصيل أكثر بكثير.**

use candle_core::{Device, Tensor}; // مكتبة candle
// ... مكتبات أخرى لل tokenization، تحميل الملفات، إلخ.

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let device = Device::cuda_if_available()?; // استخدام GPU إذا كان متاحًا، وإلا CPU

    // 1. تحميل أوزان نموذج Llama 3.2 (هذا الجزء معقد ويتطلب معالجة ملفات الأوزان)
    // let model_weights = load_llama_weights_from_files("path/to/llama_weights")?;

    // 2. تعريف بنية نموذج Llama 3.2 في Rust باستخدام candle (هذا الجزء معقد أيضًا)
    // let model = LlamaModel::new(&model_weights, device)?;

    // 3. تحميل tokenizer الخاص بـ Llama 3.2
    // let tokenizer = load_llama_tokenizer("path/to/tokenizer_config")?;

    // 4. معالجة النص المدخل (Tokenization)
    let input_text = "مرحباً بالعالم!";
    // let tokens = tokenizer.encode(input_text)?; // تحويل النص إلى رموز

    // **لتبسيط المثال جدًا، نفترض أن لدينا بالفعل رموزًا رقمية كـ Tensor:**
    let tokens_tensor = Tensor::new(&[1, 2, 3], &device)?; // رموز افتراضية كمثال

    // 5. الاستدلال (Forward pass)
    // let output_logits = model.forward(&tokens_tensor)?; // تمرير الرموز عبر النموذج

    // **لتبسيط المثال جدًا، نفترض أن لدينا لوجيتات الخرج:**
    let output_logits = Tensor::new(&[[0.1, 0.2, 0.7]], &device)?; // لوجيتات افتراضية كمثال

    // 6. اختيار الرمز التالي (Sampling) - بسيط جدًا هنا: اختيار الرمز الأعلى احتمالية
    let next_token_id = candle_nn::ops::argmax(&output_logits, -1)?;

    // 7. فك ترميز الرمز الناتج (Detokenization)
    // let generated_text = tokenizer.decode(&[next_token_id])?; // تحويل الرمز إلى نص

    println!("الرمز التالي المتوقع (كمثال): {:?}", next_token_id); // في الواقع يجب فك الترميز هنا

    Ok(())
}

Overwriting src/main.rs


In [32]:
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
error[E0432]: unresolved import `candle_core`
 --> src/main.rs:3:5
  |
3 | use candle_core::{Device, Tensor}; // مكتبة candle
  |     ^^^^^^^^^^^ use of undeclared crate or module `candle_core`

error[E0433]: failed to resolve: use of undeclared crate or module `candle_nn`
  --> src/main.rs:32:25
   |
32 |     let next_token_id = candle_nn::ops::argmax(&output_logits, -1)?;
   |                         ^^^^^^^^^ use of undeclared crate or module `candle_nn`

Some errors have detailed explanations: E0432, E0433.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `my_rust_project2` (bin "my_rust_project2") due to 2 previous errors


In [33]:
!cargo build

    Updating crates.io index
     Locking 84 packages to latest compatible versions
      Adding autocfg v1.4.0
      Adding bitflags v1.3.2
      Adding bitflags v2.8.0
      Adding bytemuck v1.21.0
      Adding bytemuck_derive v1.8.1
      Adding byteorder v1.5.0
      Adding candle-core v0.3.3 (available: v0.8.2)
      Adding candle-nn v0.3.3 (available: v0.8.2)
      Adding cfg-if v1.0.0
      Adding crc32fast v1.4.2
      Adding crossbeam-deque v0.8.6
      Adding crossbeam-epoch v0.9.18
      Adding crossbeam-utils v0.8.21
      Adding crunchy v0.2.3
      Adding dyn-stack v0.10.0
      Adding either v1.13.0
      Adding enum-as-inner v0.6.1
      Adding gemm v0.17.1
      Adding gemm-c32 v0.17.1
      Adding gemm-c64 v0.17.1
      Adding gemm-common v0.17.1
      Adding gemm-f16 v0.17.1
      Adding gemm-f32 v0.17.1
      Adding gemm-f64 v0.17.1
      Adding getrandom v0.2.15
      Adding half v2.4.1
      Adding heck v0.5.0
      Adding hermit-abi v0.3.9
      Adding itoa v1.0.

In [35]:
%%writefile src/main.rs
use std::collections::HashMap;
use std::io::{self, Write};

use candle_core::{Device, Tensor, ops}; // استيراد ops من candle_core بدلاً من candle_nn::ops

// هنا يتم تعريف "النموذج" اللغوي البسيط كقاموس (HashMap)
// في الواقع، النماذج اللغوية الحقيقية أكثر تعقيدًا بكثير
fn load_model() -> HashMap<&'static str, &'static str> {
    let mut model = HashMap::new();
    model.insert("مرحباً", "مرحباً بك!");
    model.insert("كيف حالك؟", "أنا بخير، شكراً لسؤالك.");
    model.insert("ما اسمك؟", "أنا نموذج لغوي بسيط في Rust.");
    model.insert("وداعاً", "إلى اللقاء!");
    model
}

// دالة الاستدلال الأساسية
fn infer(model: &HashMap<&'static str, &'static str>, input: &str) -> String {
    match model.get(input) {
        Some(response) => response.to_string(), // إذا وجدنا إجابة في النموذج
        None => "لا أفهم ما تقصده.".to_string(), // إذا لم نجد إجابة
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> { // إضافة Result<(), Error> للتعامل مع الأخطاء المحتملة
    let device = Device::cuda_if_available(0)?; // استخدام GPU إذا كان متاحًا، وإلا CPU، وإضافة الوسيط 0
    let model = load_model(); // تحميل "النموذج"

    println!("نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):");

    loop {
        print!("أنت: ");
        io::stdout().flush().unwrap(); // لضمان ظهور "أنت:" قبل الإدخال

        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap(); // قراءة سطر الإدخال من المستخدم

        let input = input.trim(); // إزالة المسافات البيضاء الزائدة في البداية والنهاية

        if input == "وداعاً" {
            println!("النموذج: إلى اللقاء!");
            break; // الخروج من الحلقة إذا قال المستخدم "وداعاً"
        }

        let response = infer(&model, input); // إجراء الاستدلال باستخدام النموذج والمدخل
        println!("النموذج: {}", response); // طباعة الاستجابة
    }
    Ok(()) // إضافة Ok(()) لإرجاع نتيجة ناجحة من دالة main
}

Writing src/main.rs


In [36]:
!cargo run

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
error[E0432]: unresolved import `candle_core::ops`
 --> src/main.rs:4:35
  |
4 | use candle_core::{Device, Tensor, ops}; // استيراد ops من candle_core بدلاً من candle_nn::ops
  |                                   ^^^
  |                                   |
  |                                   no `ops` in the root
  |                                   help: a similar name exists in the module: `op`
  |
  = help: consider importing one of these modules instead:
          std::ops
          candle_nn::ops

 --> src/main.rs:4:27
  |
4 | use candle_core::{Device, Tensor, ops}; // استيراد ops من candle_core بدلاً من candle_nn::ops
  |                           ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

  --> src/main.rs:26:9
   |
26 |     let device = Device::cuda_if_available(0)?; // استخدام GPU إذا كان متاحًا، وإلا CPU، وإضافة الوس...0
   |         ^^^^^^ help: if this is intentional, prefi

In [37]:
!cargo build

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
error[E0432]: unresolved import `candle_core::ops`
 --> src/main.rs:4:35
  |
4 | use candle_core::{Device, Tensor, ops}; // استيراد ops من candle_core بدلاً من candle_nn::ops
  |                                   ^^^
  |                                   |
  |                                   no `ops` in the root
  |                                   help: a similar name exists in the module: `op`
  |
  = help: consider importing one of these modules instead:
          std::ops
          candle_nn::ops

 --> src/main.rs:4:27
  |
4 | use candle_core::{Device, Tensor, ops}; // استيراد ops من candle_core بدلاً من candle_nn::ops
  |                           ^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

  --> src/main.rs:26:9
   |
26 |     let device = Device::cuda_if_available(0)?; // استخدام GPU إذا كان متاحًا، وإلا CPU، وإضافة الوس...0
   |         ^^^^^^ help: if this is intentional, prefi

In [38]:
%%writefile src/main.rs
use std::collections::HashMap;
use std::io::{self, Write};

use candle_core::{Device}; // Corrected import: Removed 'Tensor, ops' as they are not used in this simple example

// هنا يتم تعريف "النموذج" اللغوي البسيط كقاموس (HashMap)
// في الواقع، النماذج اللغوية الحقيقية أكثر تعقيدًا بكثير
fn load_model() -> HashMap<&'static str, &'static str> {
    let mut model = HashMap::new();
    model.insert("مرحباً", "مرحباً بك!");
    model.insert("كيف حالك؟", "أنا بخير، شكراً لسؤالك.");
    model.insert("ما اسمك؟", "أنا نموذج لغوي بسيط في Rust.");
    model.insert("وداعاً", "إلى اللقاء!");
    model
}

// دالة الاستدلال الأساسية
fn infer(model: &HashMap<&'static str, &'static str>, input: &str) -> String {
    match model.get(input) {
        Some(response) => response.to_string(), // إذا وجدنا إجابة في النموذج
        None => "لا أفهم ما تقصده.".to_string(), // إذا لم نجد إجابة
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> { // إضافة Result<(), Error> للتعامل مع الأخطاء المحتملة
    let device = Device::cuda_if_available(0)?; // استخدام GPU إذا كان متاحًا، وإلا CPU، وإضافة الوسيط 0 (corrected)
    let model = load_model(); // تحميل "النموذج"

    println!("نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):");

    loop {
        print!("أنت: ");
        io::stdout().flush().unwrap(); // لضمان ظهور "أنت:" قبل الإدخال

        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap(); // قراءة سطر الإدخال من المستخدم

        let input = input.trim(); // إزالة المسافات البيضاء الزائدة في البداية والنهاية

        if input == "وداعاً" {
            println!("النموذج: إلى اللقاء!");
            break; // الخروج من الحلقة إذا قال المستخدم "وداعاً"
        }

        let response = infer(&model, input); // إجراء الاستدلال باستخدام النموذج والمدخل
        println!("النموذج: {}", response); // طباعة الاستجابة
    }
    Ok(()) // إضافة Ok(()) لإرجاع نتيجة ناجحة من دالة main
}

Overwriting src/main.rs


In [40]:
%%writefile src/main.rs
use std::collections::HashMap;
use std::io::{self, Write};

use candle_core::Device; // استيراد Device فقط من candle_core

// هنا يتم تعريف "النموذج" اللغوي البسيط كقاموس (HashMap)
// في الواقع، النماذج اللغوية الحقيقية أكثر تعقيدًا بكثير
fn load_model() -> HashMap<&'static str, &'static str> {
    let mut model = HashMap::new();
    model.insert("مرحباً", "مرحباً بك!");
    model.insert("كيف حالك؟", "أنا بخير، شكراً لسؤالك.");
    model.insert("ما اسمك؟", "أنا نموذج لغوي بسيط في Rust.");
    model.insert("وداعاً", "إلى اللقاء!");
    model
}

// دالة الاستدلال الأساسية
fn infer(model: &HashMap<&'static str, &'static str>, input: &str) -> String {
    match model.get(input) {
        Some(response) => response.to_string(), // إذا وجدنا إجابة في النموذج
        None => "لا أفهم ما تقصده.".to_string(), // إذا لم نجد إجابة
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> { // إضافة Result<(), Error> للتعامل مع الأخطاء المحتملة
    // let device = Device::cuda_if_available(0)?; // تم حذف هذا السطر لأنه لم يتم استخدام المتغير 'device'
    let model = load_model(); // تحميل "النموذج"

    println!("نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):");

    loop {
        print!("أنت: ");
        io::stdout().flush().unwrap(); // لضمان ظهور "أنت:" قبل الإدخال

        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap(); // قراءة سطر الإدخال من المستخدم

        let input = input.trim(); // إزالة المسافات البيضاء الزائدة في البداية والنهاية

        if input == "وداعاً" {
            println!("النموذج: إلى اللقاء!");
            break; // الخروج من الحلقة إذا قال المستخدم "وداعاً"
        }

        let response = infer(&model, input); // إجراء الاستدلال باستخدام النموذج والمدخل
        println!("النموذج: {}", response); // طباعة الاستجابة
    }
    Ok(()) // إضافة Ok(()) لإرجاع نتيجة ناجحة من دالة main
}

Overwriting src/main.rs


@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

شغال

In [42]:
%%writefile src/main.rs
use std::collections::HashMap;
use std::io::{self, Write};

// use candle_core::Device; // تم حذف هذا السطر لأن Device لم يعد مستخدمًا

// هنا يتم تعريف "النموذج" اللغوي البسيط كقاموس (HashMap)
// في الواقع، النماذج اللغوية الحقيقية أكثر تعقيدًا بكثير
fn load_model() -> HashMap<&'static str, &'static str> {
    let mut model = HashMap::new();
    model.insert("مرحباً", "مرحباً بك!");
    model.insert("كيف حالك؟", "أنا بخير، شكراً لسؤالك.");
    model.insert("ما اسمك؟", "أنا نموذج لغوي بسيط في Rust.");
    model.insert("وداعاً", "إلى اللقاء!");
    model
}

// دالة الاستدلال الأساسية
fn infer(model: &HashMap<&'static str, &'static str>, input: &str) -> String {
    match model.get(input) {
        Some(response) => response.to_string(), // إذا وجدنا إجابة في النموذج
        None => "لا أفهم ما تقصده.".to_string(), // إذا لم نجد إجابة
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> { // إضافة Result<(), Error> للتعامل مع الأخطاء المحتملة
    // let device = Device::cuda_if_available(0)?; // تم حذف هذا السطر لأنه لم يتم استخدام المتغير 'device'
    let model = load_model(); // تحميل "النموذج"

    println!("نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):");

    loop {
        print!("أنت: ");
        io::stdout().flush().unwrap(); // لضمان ظهور "أنت:" قبل الإدخال

        let mut input = String::new();
        io::stdin().read_line(&mut input).unwrap(); // قراءة سطر الإدخال من المستخدم

        let input = input.trim(); // إزالة المسافات البيضاء الزائدة في البداية والنهاية

        if input == "وداعاً" {
            println!("النموذج: إلى اللقاء!");
            break; // الخروج من الحلقة إذا قال المستخدم "وداعاً"
        }

        let response = infer(&model, input); // إجراء الاستدلال باستخدام النموذج والمدخل
        println!("النموذج: {}", response); // طباعة الاستجابة
    }
    Ok(()) // إضافة Ok(()) لإرجاع نتيجة ناجحة من دالة main
}

Overwriting src/main.rs


In [43]:
!cargo build

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.62s


In [44]:
!cargo run

    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.06s
     Running `target/debug/my_rust_project2`
نموذج لغوي بسيط في Rust - ابدأ المحادثة (اكتب 'وداعاً' للخروج):
أنت: اهلا
النموذج: لا أفهم ما تقصده.
أنت: ^C


@@@@@@@@@@@@@@@@@@@

### ######################شغال

In [53]:
%%writefile src/main.rs
use reqwest::blocking::Client; // استخدام Client من reqwest::blocking للطلبات المتزامنة في هذا المثال البسيط
use serde_json::json;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let api_token = "XXXXXXXXXXXXXXXXX"; // **ضع توكن المصادقة الخاص بك هنا!**
    let hf_model_name = "gpt2"; // يمكنك تغيير اسم النموذج هنا (مثل "meta-llama/Llama-2-7b-chat-hf" - يتطلب توكن صالح وربما موافقة على الوصول)
    let api_url = format!("https://api-inference.huggingface.co/models/{}", hf_model_name);

    let client = Client::new();

    println!("تحدث مع نموذج Hugging Face {} (اكتب 'وداعاً' للخروج):", hf_model_name);

    loop {
        print!("أنت: ");
        std::io::Write::flush(&mut std::io::stdout())?; // تأكد من ظهور "أنت:"

        let mut input_text = String::new();
        std::io::stdin().read_line(&mut input_text)?;
        let input_text = input_text.trim();

        if input_text == "وداعاً" {
            println!("النموذج: إلى اللقاء!");
            break;
        }

        let request_body = json!({ "inputs": input_text }); // تجهيز جسم الطلب JSON

        let response = client.post(&api_url)
            .header("Authorization", format!("Bearer {}", api_token)) // إضافة توكن المصادقة
            .header("Content-Type", "application/json") // تحديد نوع المحتوى كـ JSON
            .body(request_body.to_string()) // تحويل جسم الطلب JSON إلى سلسلة وإضافته إلى الطلب
            .send()?; // إرسال الطلب

        match response.status() {
            reqwest::StatusCode::OK => {
                match response.text() { // محاولة قراءة نص الاستجابة
                    Ok(response_text) => {
                        // **هام:** استجابة API قد تكون JSON string أو JSON array، اعتمادًا على النموذج.
                        // هنا نفترض أنها JSON string بسيطة (إذا كان GPT-2، قد تحتاج إلى معالجة JSON array)
                        println!("النموذج: {}", response_text.trim_matches('"')); // طباعة الاستجابة وإزالة علامات التنصيص الزائدة
                    }
                    Err(err) => eprintln!("خطأ في قراءة نص الاستجابة من API: {}", err),
                }
            }
            _ => {
                eprintln!("خطأ من API! حالة: {}, النص: {:?}", response.status(), response.text());
            }
        }
    }

    Ok(())
}

Overwriting src/main.rs


In [54]:
!!cargo update

['\x1b\x1b    Updating\x1b crates.io index',
 '\x1b\x1b     Locking\x1b 125 packages to latest compatible versions',
 '\x1b\x1b      Adding\x1b addr2line v0.24.2',
 '\x1b\x1b      Adding\x1b adler2 v2.0.0',
 '\x1b\x1b      Adding\x1b backtrace v0.3.74',
 '\x1b\x1b      Adding\x1b base64 v0.21.7',
 '\x1b\x1b      Adding\x1b bumpalo v3.17.0',
 '\x1b\x1b    Removing\x1b bytemuck v1.21.0',
 '\x1b\x1b    Removing\x1b bytemuck_derive v1.8.1',
 '\x1b\x1b    Removing\x1b byteorder v1.5.0',
 '\x1b\x1b      Adding\x1b bytes v1.10.0',
 '\x1b\x1b    Removing\x1b candle-core v0.3.3',
 '\x1b\x1b    Removing\x1b candle-nn v0.3.3',
 '\x1b\x1b      Adding\x1b cc v1.2.13',
 '\x1b\x1b      Adding\x1b core-foundation v0.9.4',
 '\x1b\x1b      Adding\x1b core-foundation-sys v0.8.7',
 '\x1b\x1b    Removing\x1b crc32fast v1.4.2',
 '\x1b\x1b    Removing\x1b crossbeam-deque v0.8.6',
 '\x1b\x1b    Removing\x1b crossbeam-epoch v0.9.18',
 '\x1b\x1b    Removing\x1b crossbeam-utils v0.8.21',
 '\x1b\x1b    Removing\x

In [56]:
!cargo build

    Updating crates.io index
     Locking 1 package to latest compatible version
      Adding futures-io v0.3.31
  Downloaded futures-io v0.3.31
   Compiling memchr v2.7.4
   Compiling futures-task v0.3.31
   Compiling futures-io v0.3.31
   Compiling futures-util v0.3.31
   Compiling serde_json v1.0.138
   Compiling h2 v0.3.26
   Compiling hyper v0.14.32
   Compiling hyper-tls v0.5.0
   Compiling reqwest v0.11.27
   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 25.29s


In [57]:
!cargo run

    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.10s
     Running `target/debug/my_rust_project2`
تحدث مع نموذج Hugging Face gpt2 (اكتب 'وداعاً' للخروج):
أنت: اهلا
خطأ من API! حالة: 503 Service Unavailable, النص: Ok("{\"error\":\"Model openai-community/gpt2 is currently loading\",\"estimated_time\":21.923635482788086}")
أنت: ^C


In [58]:
!!cargo run

^C


['\x1b\x1b    Finished\x1b `dev` profile [unoptimized + debuginfo] target(s) in 0.22s',
 '\x1b\x1b     Running\x1b `target/debug/my_rust_project2`',
 "تحدث مع نموذج Hugging Face gpt2 (اكتب 'وداعاً' للخروج):",
 'أنت: hi',
 'خطأ من API! حالة: 503 Service Unavailable, النص: Ok("{\\"error\\":\\"Model openai-community/gpt2 is currently loading\\",\\"estimated_time\\":21.923635482788086}")',
 'أنت: ']

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@

In [ ]:
hi

In [60]:
%%writefile src/main.rs
// **تنبيه: هذا الكود هو مجرد مثال مبسط جدًا وغير قابل للتشغيل مباشرة. يحتاج إلى تفاصيل كثيرة لتشغيل نموذج حقيقي.**

use candle_core::{Device, Tensor};
// use candle_nn ...  // استيراد وحدات candle_nn اللازمة لبناء النموذج (Layers, Modules, ...)
// use ... // مكتبات لل tokenization, safetensors, ...

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let device = Device::cuda_if_available(0)?; // أو Device::Cpu إذا كنت تستخدم CPU فقط

    // 1. تحميل أوزان نموذج Llama 3.2 من ملفات safetensors
    // let model_weights = load_model_weights_from_safetensors("path/to/model.safetensors")?;

    // 2. تعريف بنية نموذج Llama 3.2 في Rust باستخدام candle_nn
    // let model = LlamaForCausalLM::load(&model_weights, &config, device)?;

    // 3. تحميل tokenizer الخاص بـ Llama 3.2
    // let tokenizer = load_tokenizer("path/to/tokenizer.json")?;

    // حلقة المحادثة (مشابهة للمثال السابق مع API)
    loop {
        // ... قراءة النص المدخل من المستخدم ...

        // 4. Tokenization للنص المدخل
        // let tokens = tokenizer.encode(input_text)?;
        // let tokens_tensor = Tensor::new(tokens, &device)?;

        // 5. الاستدلال (Forward pass)
        // let output_logits = model.forward(&tokens_tensor)?;

        // 6. Sampling لاختيار الرمز التالي
        // let next_token_id = sample_next_token(&output_logits)?;

        // 7. Detokenization للرمز الناتج
        // let generated_text = tokenizer.decode(&[next_token_id])?;

        // ... طباعة الاستجابة ...
    }

    Ok(())
}

Writing src/main.rs


In [61]:
!cargo build

   Compiling my_rust_project2 v0.1.0 (/content/my_rust_project/my_rust_project2)
error[E0432]: unresolved import `candle_core`
 --> src/main.rs:3:5
  |
3 | use candle_core::{Device, Tensor};
  |     ^^^^^^^^^^^ use of undeclared crate or module `candle_core`

  --> src/main.rs:39:5
   |
20 | /     loop {
21 | |         // ... قراءة النص المدخل من المستخدم ...
22 | |
23 | |         // 4. Tokenization للنص المدخل
...  |
36 | |         // ... طباعة الاستجابة ...
37 | |     }
   | |_____- any code following this expression is unreachable
38 |
39 |       Ok(())
   |       ^^^^^^ unreachable expression
   |
   = note: `#[warn(unreachable_code)]` on by default

For more information about this error, try `rustc --explain E0432`.
error: could not compile `my_rust_project2` (bin "my_rust_project2") due to 1 previous error; 1 warning emitted


https://github.com/Codehackerone/storyforge/blob/main/rust-server/Cargo.toml

In [ ]:
[package]
name = "rust-server"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
tonic = "0.7.1"
tokio = { version = "1.18.0", features = ["macros", "rt-multi-thread"] }
prost = "0.10.1"

[build-dependencies]
tonic-build = "0.7.2"
tonic-reflection = "0.4.0"

[[bin]]
name = "python-client"
path = "src/python-client.rs"

[[bin]]
name="cli"
path="src/cli.rs"

In [62]:
%%writefile src/main.rs
use std::{env, path::PathBuf};

fn main() {
    let proto_file = "textgen.proto";

    tonic_build::configure()
        .build_server(true)
        .out_dir("./src")
        .compile(&[proto_file], &["."])
        .unwrap_or_else(|e| panic!("protobuf compile error: {}", e));

    println!("cargo:rerun-if-changed={}", proto_file);
}

Overwriting src/main.rs


In [64]:
!cargo build

    Updating crates.io index
     Locking 52 packages to latest compatible versions
      Adding aho-corasick v1.1.3
      Adding anyhow v1.0.95
      Adding async-stream v0.3.6
      Adding async-stream-impl v0.3.6
      Adding async-trait v0.1.86
      Adding axum v0.5.17
      Adding axum-core v0.2.9
 Downgrading base64 v0.21.7 -> v0.13.1
      Adding byteorder v1.5.0
      Adding cmake v0.1.54
      Adding either v1.13.0
      Adding fixedbitset v0.4.2
      Adding getrandom v0.2.15
      Adding hashbrown v0.12.3
      Adding heck v0.4.1
      Adding home v0.5.11
      Adding http-range-header v0.3.1
      Adding hyper-timeout v0.4.1
      Adding indexmap v1.9.3
      Adding itertools v0.10.5
      Adding lazy_static v1.5.0
      Adding matchit v0.5.0
      Adding multimap v0.8.3
      Adding petgraph v0.6.5
      Adding pin-project v1.1.9
      Adding pin-project-internal v1.1.9
      Adding ppv-lite86 v0.2.20
      Adding prettyplease v0.1.25
      Adding prost v0.10.4 (available

In [ ]:
!cargo run

In [68]:
%cd /content

/content


https://github.com/rustformers/llm#running





https://docs.rs/llm/latest/src/llm/builder.rs.html










https://github.com/rustformers/llmhttps://github.com/rustformers/llmhttps://github.com/rustformers/llmhttps://github.com/rustformers/llmhttps://github.com/rustformers/llmhttps://github.com/rustformers/llm

### $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$

In [69]:
!git clone https://github.com/rustformers/llm.git
!nix build

Cloning into 'llm'...
remote: Enumerating objects: 10402, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 10402 (delta 0), reused 1 (delta 0), pack-reused 10399 (from 1)
Receiving objects: 100% (10402/10402), 6.49 MiB | 19.59 MiB/s, done.
Resolving deltas: 100% (5037/5037), done.
/bin/bash: line 1: nix: command not found


In [70]:
%cd llm

/content/llm


In [71]:
!nix build

/bin/bash: line 1: nix: command not found


https://stackoverflow.com/questions/76675438/how-to-compile-the-llm-library-with-nix

In [72]:
!nix develop -i
!cargo build --release

/bin/bash: line 1: nix: command not found
    Updating crates.io index
  Downloaded addr2line v0.21.0
  Downloaded zstd-safe v6.0.5+zstd.1.5.4
  Downloaded aes v0.8.3
  Downloaded zstd v0.11.2+zstd.1.5.2
  Downloaded itoa v1.0.6
  Downloaded utf8parse v0.2.1
  Downloaded indicatif v0.16.2
  Downloaded tracing-subscriber v0.3.17
  Downloaded partial_sort v0.2.0
  Downloaded anstyle-parse v0.2.1
  Downloaded rustyline-derive v0.8.0
  Downloaded aho-corasick v0.7.20
  Downloaded tracing-appender v0.2.2
  Downloaded memoffset v0.9.0
  Downloaded termcolor v1.2.0
  Downloaded percent-encoding v2.3.0
  Downloaded adler v1.0.2
  Downloaded time-macros v0.2.9
  Downloaded tracing-log v0.1.3
  Downloaded memmap2 v0.5.10
  Downloaded tracing-attributes v0.1.26
  Downloaded tracing-core v0.1.31
  Downloaded tracing v0.1.37
  Downloaded unicode-normalization-alignments v0.1.12
  Downloaded serde_json v1.0.99
  Downloaded reqwest v0.11.18
  Downloaded rayon v1.7.0
  Downloaded syn v2.0.22
  Downloa

https://github.com/rustformers/llm#running

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

In [73]:
!cargo install --git https://github.com/rustformers/llm llm-cli

    Updating git repository `https://github.com/rustformers/llm`
    Updating git submodule `https://github.com/ggerganov/llama.cpp`
  Installing llm-cli v0.2.0-dev (https://github.com/rustformers/llm#b11ffb14)
    Updating crates.io index
     Locking 324 packages to latest compatible versions
      Adding backtrace v0.3.71 (available: v0.3.74)
      Adding env_logger v0.10.2 (available: v0.11.6)
      Adding memmap2 v0.5.10 (available: v0.9.5)
      Adding partial_sort v0.2.0 (available: v1.0.0)
      Adding rand v0.8.5 (available: v0.9.0)
      Adding rustyline v11.0.0 (available: v15.0.0)
      Adding thiserror v1.0.69 (available: v2.0.11)
      Adding tokenizers v0.13.4 (available: v0.21.0)
      Adding zstd v0.12.4 (available: v0.13.2)
  Downloaded anstream v0.6.18
  Downloaded bytesize v1.3.2
  Downloaded anstyle v1.0.10
  Downloaded anstyle-parse v0.2.6
  Downloaded anstyle-query v1.1.2
  Downloaded adler2 v2.0.0
  Downloaded time-core v0.1.2
  Downloaded monostate-impl v0.1.13

In [75]:
%cd /content
!git clone https://huggingface.co/togethercomputer/RedPajama-INCITE-Base-3B-v1

/content
Cloning into 'RedPajama-INCITE-Base-3B-v1'...
remote: Enumerating objects: 75, done.
remote: Total 75 (delta 0), reused 0 (delta 0), pack-reused 75 (from 1)
Unpacking objects: 100% (75/75), 601.38 KiB | 4.97 MiB/s, done.
You can inspect what was checked out with 'git status'
and retry with 'git restore --source=HEAD :/'


Exiting because of "interrupt" signal.
^C


In [76]:
%cd /content/RedPajama-INCITE-Base-3B-v1
!wget https://huggingface.co/togethercomputer/RedPajama-INCITE-Base-3B-v1/resolve/main/pytorch_model.bin

/content/RedPajama-INCITE-Base-3B-v1
--2025-02-13 03:39:11--  https://huggingface.co/togethercomputer/RedPajama-INCITE-Base-3B-v1/resolve/main/pytorch_model.bin
Resolving huggingface.co (huggingface.co)... 18.172.134.4, 18.172.134.88, 18.172.134.24, ...
Connecting to huggingface.co (huggingface.co)|18.172.134.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cdn-lfs.hf.co/repos/46/28/4628639db4a82781b4ff9dd35e306d62dabf40d7d08628e594c71fc3220cb0b2/844708c42241de49775b872663ad58422a35b8e5a7d50b1f808408db8c1fb84b?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27pytorch_model.bin%3B+filename%3D%22pytorch_model.bin%22%3B&response-content-type=application%2Foctet-stream&Expires=1739421551&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTczOTQyMTU1MX19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5oZi5jby9yZXBvcy80Ni8yOC80NjI4NjM5ZGI0YTgyNzgxYjRmZjlkZDM1ZTMwNmQ2MmRhYmY0MGQ3ZDA4NjI4ZTU5NGM3MWZjMzIyMGNiMGIyLzg0

https://docs.rs/llm/latest/src/llm/chat/mod.rs.html#1-106

https://docs.rs/llm/latest/src/llm/builder.rs.html

https://huggingface.co/rustformers/redpajama-3b-ggml/blob/main/README.md?code=true#L25

In [79]:
%cd /content
!wget https://huggingface.co/rustformers/redpajama-3b-ggml/resolve/main/RedPajama-INCITE-Base-3B-v1-q4_0.bin

/content
--2025-02-13 03:44:54--  https://huggingface.co/rustformers/redpajama-3b-ggml/resolve/main/RedPajama-INCITE-Base-3B-v1-q4_0.bin
Resolving huggingface.co (huggingface.co)... 18.172.134.124, 18.172.134.24, 18.172.134.88, ...
Connecting to huggingface.co (huggingface.co)|18.172.134.124|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://cdn-lfs.hf.co/repos/1e/6b/1e6b7b71c8679230d45d3d1f5cf71effa2052c1608e85061cb8a87dce0513282/549b4f8d5f153aacdd635ff999d16f63ed0e7b8458ad1a1a41ac235a89e4582d?response-content-disposition=inline%3B+filename*%3DUTF-8%27%27RedPajama-INCITE-Base-3B-v1-q4_0.bin%3B+filename%3D%22RedPajama-INCITE-Base-3B-v1-q4_0.bin%22%3B&response-content-type=application%2Foctet-stream&Expires=1739421894&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTczOTQyMTg5NH19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy5oZi5jby9yZXBvcy8xZS82Yi8xZTZiN2I3MWM4Njc5MjMwZDQ1ZDNkMWY1Y2Y3MWVmZmEyMDUyYzE2MDhlODUwNjFjYjhhOD

In [80]:
!llm infer -a gptneox -m /content/RedPajama-INCITE-Base-3B-v1-q4_0.bin -p "Rust is a cool programming language because" -r togethercomputer/RedPajama-INCITE-Base-3B-v1

✓ Loaded 388 tensors (1.6 GB) after 2182ms
Rust is a cool programming language because it compiles to C code which has been proven for decades. It’s also cross platform so you can build on Windows or Linux without having both runtimes installed, and the libraries are all open source projects that people have contributed back into Rust including some of them from other languages in use by Google today such as LLVM/Clang compiler (similar to Clojure) which compiles C code for Mac OS X.
You can also take advantage of JITs like Lazy Byte’s Just-In Time Debugger, or Apple’s Swift Compiler with their respective libraries and build rust on the fly without compiling from command line using just system resources in a few minutes to hours depending upon your CPU/OS type (and how much RAM it has). Rust isn't meant for massive games though.
There is an OpenGL implementation that you can use if want's graphics support which seems like more of what we need than native code at the moment and there ar

✓ Loaded 388 tensors (1.6 GB) after 2182ms
Rust is a cool programming language because it compiles to C code which has been proven for decades. It’s also cross platform so you can build on Windows or Linux without having both runtimes installed, and the libraries are all open source projects that people have contributed back into Rust including some of them from other languages in use by Google today such as LLVM/Clang compiler (similar to Clojure) which compiles C code for Mac OS X.
You can also take advantage of JITs like Lazy Byte’s Just-In Time Debugger, or Apple’s Swift Compiler with their respective libraries and build rust on the fly without compiling from command line using just system resources in a few minutes to hours depending upon your CPU/OS type (and how much RAM it has). Rust isn't meant for massive games though.
There is an OpenGL implementation that you can use if want's graphics support which seems like more of what we need than native code at the moment and there are other libraries such as JSON from C++, Irimagi’s WebRender (like Cocoa Touch but with JavaScript), Rust bindings to popular APIs for Ruby/Python etc.
There have been various studies done recently showing that a large portion or all developers would like open source software in their company environment today despite being somewhat skeptical about using an unknown language and not having enough time on hand, especially when there are other tools available such as JetBrain’s JRuby which compiles Java to Ruby for Mac OS X.
And now comes the problem of Rust taking over C where it was a weakling before - developers that wrote code in non-C languages like Python and JavaScript would have had no choice but just migrate their libraries into using native platforms such as .NET with its excellent development environment, not really having to care about compiler performance.
So they’re still on the fence for now until someone comes up a way around this - it seems most of them are considering migrating projects over time and some will stick there while others just go full speed ahead onto Rust which is what I'm planning too after seeing how much love JRuby has received from Java developers, especially during conferences.
What we really need for C to be replaced by something like Swift with a compiler that compiles native code but can also take advantage of other languages such as Python/JavaScript and not just Ruby or even Javascript itself if it's going full steam ahead - there are lots more than those I'm sure, so having someone come up around then would have been the best choice in my opinion.
JRuby has already got a huge following especially among Java developers who don't care what language they use for development because its performance and features make C obsolete today; but now that Rust is taking over even more of them will migrate to it or at least give serious consideration if not yet commit time into making their existing libraries convertible.
This year we’ll be seeing^C

✓ تم تحميل 388 موترًا (1.6 جيجابايت) بعد 2182 مللي ثانية
Rust هي لغة برمجة رائعة لأنها تُجمَّع إلى كود C الذي أثبت كفاءته لعقود من الزمن. كما أنها متعددة المنصات بحيث يمكنك البناء على Windows أو Linux دون الحاجة إلى تثبيت كلا بيئتي التشغيل، والمكتبات كلها عبارة عن مشاريع مفتوحة المصدر ساهم بها الناس في Rust بما في ذلك بعضها من لغات أخرى تستخدمها Google اليوم مثل مُجمِّع LLVM/Clang (مشابه لـ Clojure) الذي يُجمِّع كود C لنظام التشغيل Mac OS X.
يمكنك أيضًا الاستفادة من JITs مثل Just-In Time Debugger من Lazy Byte، أو Swift Compiler من Apple مع مكتباتها الخاصة وبناء Rust أثناء التنقل دون التجميع من سطر الأوامر باستخدام موارد النظام فقط في بضع دقائق إلى ساعات اعتمادًا على نوع وحدة المعالجة المركزية/نظام التشغيل (وكمية ذاكرة الوصول العشوائي التي يحتوي عليها). ومع ذلك، فإن Rust غير مخصص للألعاب الضخمة.
هناك تطبيق OpenGL يمكنك استخدامه إذا كنت تريد دعمًا للرسومات، ويبدو أنه أكثر مما نحتاجه من التعليمات البرمجية الأصلية في الوقت الحالي، وهناك مكتبات أخرى مثل JSON من C++، وWebRender من Irimagi (مثل Cocoa Touch ولكن مع JavaScript)، وروابط Rust لواجهات برمجة التطبيقات الشائعة لـ Ruby/Python وما إلى ذلك.
لقد أجريت دراسات مختلفة مؤخرًا أظهرت أن جزءًا كبيرًا أو جميع المطورين يرغبون في استخدام برامج مفتوحة المصدر في بيئة شركتهم اليوم على الرغم من تشككهم إلى حد ما في استخدام لغة غير معروفة وعدم وجود وقت كافٍ في متناول اليد، خاصةً عندما تتوفر أدوات أخرى مثل JRuby من JetBrain والتي تقوم بتجميع Java إلى Ruby لنظام التشغيل Mac OS X.
والآن تأتي مشكلة Rust التي تغلبت على C حيث كانت ضعيفة من قبل - لم يكن أمام المطورين الذين كتبوا التعليمات البرمجية بلغات غير C مثل Python وJavaScript أي خيار سوى نقل مكتباتهم إلى استخدام منصات أصلية مثل .NET مع بيئة التطوير الممتازة الخاصة بها، دون الاضطرار حقًا إلى الاهتمام بأداء المترجم.
لذا فهم ما زالوا مترددين في الوقت الحالي حتى يأتي شخص ما بطريقة للالتفاف على هذا - يبدو أن معظمهم يفكرون في نقل المشاريع بمرور الوقت وسيلتزم البعض بذلك بينما يمضي آخرون بأقصى سرعة نحو Rust وهو ما أخطط له أيضًا بعد رؤية مقدار الحب الذي تلقاه JRuby من مطوري Java، وخاصة أثناء المؤتمرات.
ما نحتاجه حقًا لاستبدال C بشيء مثل Swift مع مُجمِّع يقوم بتجميع الكود الأصلي ولكن يمكنه أيضًا الاستفادة من لغات أخرى مثل Python / JavaScript وليس فقط Ruby أو حتى Javascript نفسها إذا كان يعمل بكامل طاقته - هناك الكثير غير ذلك أنا متأكد، لذلك فإن وجود شخص ما سيأتي بعد ذلك سيكون الخيار الأفضل في رأيي.
لقد حصل JRuby بالفعل على عدد كبير من المتابعين خاصة بين مطوري Java الذين لا يهتمون باللغة التي يستخدمونها للتطوير لأن أداءها وميزاتها تجعل C عتيقة اليوم؛ ولكن الآن بعد أن استولى Rust على المزيد منهم سيهاجرون إليه أو على الأقل يفكرون بجدية إن لم يكن قد خصصوا الوقت لجعل مكتباتهم الحالية قابلة للتحويل.
سنرى هذا العام ^C

In [82]:
!llm infer -a gptneox -m /content/RedPajama-INCITE-Base-3B-v1-q4_0.bin -p "Who is Napoleon Bonaparte?" -r togethercomputer/RedPajama-INCITE-Base-3B-v1 --num-predict 10

✓ Loaded 388 tensors (1.6 GB) after 2961ms
Who is Napoleon Bonaparte?
The French general who would become one of history


https://huggingface.co/rustformers

In [81]:
!llm infer -h

Use a model to infer the next tokens in a sequence, and exit

Usage: llm infer [OPTIONS] --model-path <MODEL_PATH>

Options:
  -m, --model-path <MODEL_PATH>
          Where to load the model from
  -a, --model-architecture <MODEL_ARCHITECTURE>
          The model architecture to use. Will attempt to guess if not specified
  -v, --tokenizer-path <TOKENIZER_PATH>
          Local path to Hugging Face tokenizer file
  -r, --tokenizer-repository <TOKENIZER_REPOSITORY>
          Remote Hugging Face repository containing a tokenizer
      --num-ctx-tokens <NUM_CTX_TOKENS>
          Sets the size of the context (in tokens). Allows feeding longer prompts. Note that this affects memory [default: 2048]
      --no-mmap
          Don't use mmap to load the model
      --lora-paths [<LORA_PATHS>...]
          LoRA adapter to use for the model
      --gpu-layers <GPU_LAYERS>
          Number of layers to run on the GPU. If not specified, all layers will be run on the GPU
      --rope-freq-base <ROPE_

Use a model to infer the next tokens in a sequence, and exit

Usage: llm infer [OPTIONS] --model-path <MODEL_PATH>

Options:
  -m, --model-path <MODEL_PATH>
          Where to load the model from
  -a, --model-architecture <MODEL_ARCHITECTURE>
          The model architecture to use. Will attempt to guess if not specified
  -v, --tokenizer-path <TOKENIZER_PATH>
          Local path to Hugging Face tokenizer file
  -r, --tokenizer-repository <TOKENIZER_REPOSITORY>
          Remote Hugging Face repository containing a tokenizer
      --num-ctx-tokens <NUM_CTX_TOKENS>
          Sets the size of the context (in tokens). Allows feeding longer prompts. Note that this affects memory [default: 2048]
      --no-mmap
          Don't use mmap to load the model
      --lora-paths [<LORA_PATHS>...]
          LoRA adapter to use for the model
      --gpu-layers <GPU_LAYERS>
          Number of layers to run on the GPU. If not specified, all layers will be run on the GPU
      --rope-freq-base <ROPE_FREQ_BASE>
          
      --rope-freq-scale <ROPE_FREQ_SCALE>
          
  -f, --prompt-file <PROMPT_FILE>
          A file to read the prompt from
  -t, --num-threads <NUM_THREADS>
          Sets the number of threads to use
  -n, --num-predict <NUM_PREDICT>
          Sets how many tokens to predict
      --batch-size <BATCH_SIZE>
          How many tokens from the prompt at a time to feed the network. Does not affect generation [default: 8]
  -s, --sampler <SAMPLER_OPTIONS>
          Configure sampler settings using a string in the format: sampler_name:key1=value1:key2=value2
          To configure multiple samplers at once, separate the sampler configuration strings with space or '/' (forward slash).
          NOTE: Mirostat samplers are incompatible with top-p, top-k, locally typical and tail free samplers.
          TIPS:
            1. Sampler options aren't required. For example "mirostat1" will enable Mirostat 1 with its default options.
            2. It's possible to specify partial option names, as long as they are unambiguous.
            3. Underscore and dash are ignored in sampler names, so "top-p" is the same as "topp" or "top_p".
      --seed <SEED>
          Specifies the seed to use during sampling. Note that, depending on hardware, the same seed may lead to different results on two separate machines
      --no-float16
          Use 32-bit floats for model memory key and value. Not recommended: doubles size without a measurable quality increase. Ignored when restoring from the cache
      --token-bias <TOKEN_BIAS>
          A comma separated list of token biases. The list should be in the format "TID=BIAS,TID=BIAS" where TID is an integer token ID and BIAS is a floating point number. For example, "1=-1.0,2=-1.0" sets the bias for token IDs 1 (start of document) and 2 (end of document) to -1.0 which effectively disables the model from generating responses containing those token IDs
      --ignore-eos
          Prevent the end of stream (EOS/EOD) token from being generated. This will allow the model to generate text until it runs out of context space
      --use-gpu
          Whether to use GPU acceleration when available
  -p, --prompt <PROMPT>
          The prompt to feed the generator
      --hide-prompt
          Hide the prompt in the generation
      --load-session <LOAD_SESSION>
          Loads a saved inference session from the given path, previously saved using `--save-session`
      --save-session <SAVE_SESSION>
          Saves an inference session at the given path. The same session can then be loaded from disk using `--load-session`
      --persist-session <PERSIST_SESSION>
          Loads an inference session from the given path if present, and then saves the result to the same path after inference is completed
      --stats
          Output statistics about the time taken to perform inference, among other things
  -h, --help
          Print help (see more with '--help')

https://github.com/sekmet/llm-rust

https://github.com/jondot/awesome-rust-llm

$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$